src/shogun/util/traits.h

```diff
@@ -184,6 +185,29 @@ namespace shogun
 
 		template<uint32_t idx, typename ...Ts>
 		using get_variant_type_t = typename get_variant_type<idx, Ts...>::type;
+
+		template<typename>
+		struct is_atomic: std::false_type {};
+
+		template <typename T>
+		struct is_atomic<std::atomic<T>> : std::true_type
+		{
+		};
+
+		template <typename T>
+		struct atomic_type {};
```

Comment:
replaces std::atomic<T>::value_type, because it is only available in very recent compilers (gcc-9 and clang trunk)